### PR TITLE
postgresql_pg_hba: bulk rule editing

### LINKED ISSUE #297: feature request: bulk option for pg_hba module
### feature idea for module postgresql_pg_hba: bulk rule editing

Some of my pg_hba.conf have over 700 managed rules, which takes some time and memory to loop over (repeated parsing, editing and writing). My current usecase looks like this:
```yaml
- name: manage pg_hba.conf
  community.postgresql.postgresql_pg_hba:
    create: yes
    dest: "{{ data_dir }}/pg_hba.conf"
    backup: yes
    backup_file: "{{ data_dir }}/pg_hba.conf.ansible_backup"
    keep_comments_at_rules: true
   # not used here: attributes, group, mode, owner, selevel, serole, setype, seuser, unsafe_writes

    contype: "{{ entry.host }}"
    method: "{{ entry.method }}"
    options: "{{ entry.options }}"
    address: "{{ entry.address }}"
    netmask: "{{ entry.netmask }}"
    databases: "{{ entry.databases }}"
    state: "{{ entry.state }}"
    users: "{{ entry.users }}"
    comment: "{{ entry.comment }}"
  loop: "{{ desired_state.pghba|default([]) }}"
  loop_control:
    loop_var: entry
```

I would like to optimize that by adding/removing multiple rules with a single call to the module, which could look like this:
```yaml
- name: manage pg_hba.conf
  community.postgresql.postgresql_pg_hba:
    create: yes
    dest: "{{ data_dir }}/pg_hba.conf"
    backup: yes
    backup_file: "{{ data_dir }}/pg_hba.conf.ansible_backup"
    keep_comments_at_rules: true
   # not used here: attributes, group, mode, owner, selevel, serole, setype, seuser, unsafe_writes

    bulk: "{{ desired_state.pghba|default([]) }}"
```

The new `bulk` argument (or `rules`?) would take a list of objects, each specifying a rule using the existing arguments:
```
address
comment
contype
databases
method
netmask
options
state
users
```
If one object is missing one of these, the default is taken, if possible.

When the arguments which can be put into the `bulk` argument are set normally (example below), they could either
1) connflict with the `bulk` argument
2) be appended to the `bulk` argument
3) be treated as defaults for missing values in the `bulk` argument

```yaml
- name: manage pg_hba.conf
  community.postgresql.postgresql_pg_hba:
    contype: host
    address: "2001:db8:1::/64"
    databases: db1
    users: user1
    state: absent

    bulk:
    - contype: host
      address: "2001:db8:2::/64"
      databases: db2
      users: user2
# option 1 would let the module fail

# option 2 would be equivalent to:
- name: manage pg_hba.conf
  community.postgresql.postgresql_pg_hba:
    bulk:
    - contype: host
      address: "2001:db8:1::/64"
      databases: db1
      users: user1
      state: absent
    - contype: host
      address: "2001:db8:2::/64"
      databases: db2
      users: user2
      # implicit "state: present"

# option 3 would be equivalent to:
- name: manage pg_hba.conf
  community.postgresql.postgresql_pg_hba:
    bulk:
    - contype: host
      address: "2001:db8:2::/64"
      databases: db2
      users: user2
      state: absent  # normal "state" argument used as default
```

### PR DESCRIPTION
**Pull request by betanummeric**
feature PR, fixes #297

#### argument `rules`
A list of objects, specifying a rule for the pg_hba.conf. Each object can have the following keys (the "rule-specific arguments"), which are treated the same as if they were arguments of this module:
 - `address`
 - `comment`
 - `contype`
 - `databases`
 - `method`
 - `netmask`
 - `options`
 - `state`
 - `users`
See `rules_behavior` for what happens when one of these keys missing in a `rules` item.

#### argument `rules_behavior`
String, choices: `conflict` (default), `combine`.

When `conflict`, and `rules` is specified along with at least one normal rule-specific argument, the module fails. This should prevent accidental ambiguous operations. If a `rules` item is lacking some rule-specific argument, the respective module default is taken.

When `combine`, the normal rule-specific arguments are treated as fallback values for the content of `rules`. (They will not add or remove a role on their own.) When a value is neither specified in a `rules` item nor in the normal rule-specific argument, the module default is taken.

example:
```yaml
- name: rules with custom defaults
  community.postgresql.postgresql_pg_hba:
    dest: /tmp/pg_hba.conf
    rules_behavior: combine
    databases: db1
    address: 2001:db8::a/64
    rules:
      - users: user1    
      - users: user2
        address: 2001:db8::b/64

# is equivalent to
- name: rules with custom defaults
  community.postgresql.postgresql_pg_hba:
    dest: /tmp/pg_hba.conf
    rules_behavior: combine
    rules:
      - users: user1    
        databases: db1
        address: 2001:db8::a/64
      - users: user2
        databases: db1
        address: 2001:db8::b/64
```

#### argument `overwrite`
Boolean, default: false. When true, all pre-existing rules will be removed from the pg_hba.conf as if they where specified with `state: absent`.

### CODE REVIEW

**Comment by hunleyd** [comment]
@betanummeric any updates on the remaining failures ?

**Comment by betanummeric** [comment]
The remaining error message occurs [here](https://github.com/betanummeric/community.postgresql/blob/pghba-bulk-editing/tests/integration/targets/postgresql_pg_hba/tasks/postgresql_pg_hba_bulk_rules.yml#L79) and says:
```
The conditional check 'result.results|selectattr('msg', 'in', 'conflict')|length == 0' failed. The error was: no test named 'in'
```
The test runs in docker image `quay.io/ansible/centos7-test-container:2.0.2` with python 2.7.5 and jinja 2.7.2, but the `in`-test was added in jinja 2.10. Even 
```
result.results|map(attribute='msg')|map('truncate',8,True,'')|select('equalto','conflict')|length == 0
```
would not work because `equalto` was added in jina 2.8. I tried some ways but I see no way to perform this check with jinja in that version.
I would like to ignore or skip that check with that version. In the end, it's only testing that the error messages look sensible.

**Comment by Andersson007** [comment]
@betanummeric 

>I would like to ignore or skip that check with that version. In the end, it's only testing that the error messages look sensible.

Skipping or ignoring it with that version sounds OK to me

**Comment by betanummeric** [comment]
The tests are green, this is ready for review.

File: plugins/modules/postgresql_pg_hba.py
Comment (Andersson007) [inline plugins/modules/postgresql_pg_hba.py:110]:
```suggestion
      - A list of objects, specifying rules for the pg_hba.conf. Use this to manage multiple rules at once.
      - "Each object can have the following keys (the 'rule-specific arguments'), which are treated the same as if they were arguments of this module:"
```
For better readability

**Review by Andersson007 — COMMENTED** [review]
@betanummeric thanks for working on this!

I left several general things.
I'm not a user of the module, so would be nice to hear users' / other maintainers' opinions.

@hunleyd @jchancojr do you use the module or do you know someone who uses it and can take a look?

**Comment by hunleyd** [comment]
> @hunleyd @jchancojr do you use the module or do you know someone who uses it and can take a look?

myself @pgguru and @keithf4 use the module at $dayjob. we can take a look

File: plugins/modules/postgresql_pg_hba.py
Comment (betanummeric) [reply plugins/modules/postgresql_pg_hba.py:777]: When `rules_behavior` is specified without `rules`, it is currently ignored. I see no way to tell ansible "arg1 must be unset if arg2 is unset" using the linked module options. I could add the error in the module code if you want that.

File: plugins/modules/postgresql_pg_hba.py
Comment (Andersson007) [inline plugins/modules/postgresql_pg_hba.py:118]:
```suggestion
      - If C(combine), the normal rule-specific arguments are not defining a rule, but are used as defaults for the arguments in the I(rules) argument.
      - Is used only when I(rules) is specified, ignored otherwise.
```

File: plugins/modules/postgresql_pg_hba.py
Comment (Andersson007) [reply plugins/modules/postgresql_pg_hba.py:777]: @betanummeric OK, how about https://github.com/ansible-collections/community.postgresql/pull/303#discussion_r915531449 then?

**Comment by Andersson007** [comment]
> > @hunleyd @jchancojr do you use the module or do you know someone who uses it and can take a look?
> 
> myself @pgguru and @keithf4 use the module at $dayjob. we can take a look

Great, thanks! We'll be waiting for you feedback

**Review by Andersson007 — APPROVED** [review]

**Comment by hunleyd** [comment]
does this need backported once merged @Andersson007 ?

**Comment by sebasmannem** [comment]
Awesome work guys. 
Thanks.

**Comment by Andersson007** [comment]
> does this need backported once merged @Andersson007 ?

@hunleyd no, it doesn't, thanks for merging!

**Comment by Andersson007** [comment]
@betanummeric thanks for the contribution!

**Pull request merged**
